net: sockets: tls: Check accepted context is not NULL

lbise opened the pull request:
Fixes a NULL pointer dereferencement when closing a listening TLS socket. In that case k_fifo_get will be unblocked and return a NULL child which will then be dereferenced in the code below.

lbise committed:
net: sockets: tls: Check accepted context is not NULL

When closing a listening socket the functions waiting on the
FIFO will be unblocked this will result in receiving a NULL child
context. If that is the case return an error instead of carrying on.
Return the same error code (EINVAL) that is returned on Linux when
calling shutdown on a blocked accept call.

Signed-off-by: Léonard Bise <leonard.bise@gmail.com>

jukkar commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -1277,6 +1277,11 @@ int ztls_accept_ctx(struct net_context *parent, struct sockaddr *addr,
 	child = k_fifo_get(&parent->accept_q, K_FOREVER);
+	if (child == NULL) {
+		z_free_fd(fd);
+		errno = EAGAIN;
```
I just wonder if this is proper errno value when the listening socket is closed. In the calling app, a new socket needs to be created in that case. Would `EBADF` be more proper, at least the `accept()` manual page says `EBADF  sockfd is not an open file descriptor.`, what do you think?

lbise replied:
EBADF makes sense I guess. I used the same error returned by sockets.c but this takes non-blocking sockets in mind I guess.

@rlubos By the way side question but are there plans to support async TLS sockets in the future?

rlubos replied:
I agree, since `EAGAIN` should only be used with non-blocking sockets, it might be confusing to receive it on a blocking one, plus ambiguous with non-blocking sockets. `EBADF` sounds like a good alternative.

rlubos replied:
> By the way side question but are there plans to support async TLS sockets in the future?

If you mean async handshake, not on my roadmap currently I'm afriad. As for regular `send`/`recv` it should work I believe (unless some bugs slipped in over time).

pfalcon replied:
EAGAIN has very specific meaning (non-blocking operation was requested, and operation can't complete without blocking), and shouldn't be used for anything else. Even looking at the 2 last lines makes it clear that EAGAIN is wrong choice here: you can't really "free" fd and return EAGAIN, the whole idea of EAGAIN that you're supposed to call the same (or another) function on that fd again, so you can't close or free it. (EAGAIN isn't even an "error" per se, it's "transient condition").

lbise replied:
I'll change it but just FYI sockets.c : zsock_accept_ctx uses the exact same code ;)

lbise replied:
> > By the way side question but are there plans to support async TLS sockets in the future?
> 
> If you mean async handshake, not on my roadmap currently I'm afriad. As for regular `send`/`recv` it should work I believe (unless some bugs slipped in over time).

I meant more of having accept be non-blocking, currently it blocks so you cannot really have a single loop do the accept / send / recv thing in parallel if I'm not mistaken. Does what I say makes sense?

pfalcon replied:
> If you mean async handshake, not on my roadmap currently I'm afriad.

To properly implement async TLS handshake, you really need to do it in "userspace", instead of layering-violation implementation you guys have now, contrary to warnings of yours truly ;-). The problem, as you imagine, lies in the fact of not enough selectivity in non-blocking return codes in POSIX stream (fd) model. To support generalized streams like TLS, you should be prepared to a situation when write() requires blocking for read operations, and vice versa. POSIX offers just EAGAIN to convey that write() needs blocking for write(), or read() - blocking for read(), no return code for the "cross" situation.

(And I can very well imagine that you'll persist in your wrongdoing and will introduce EAGAIN_RD and EAGAIN_WR, which will only further violate POSIX API contract.)

pfalcon replied:
@pfalcon wrote:

> Even looking at the 2 last lines makes it clear that EAGAIN is wrong choice here: you can't really "free" fd and return EAGAIN,

Ok, let's call this ~~brain f%rt~~ uncareful looking: here you free the newly allocated fd, so the above tirade doesn't apply, sorry for the noise.

pfalcon replied:
@lbise:

> I'll change it but just FYI sockets.c : zsock_accept_ctx uses the exact same code ;)

Ok, gotcha. @lbise, can I please ask you to investigate this issue to get to the bottom of it - code up a Linux sample, check what it does, and do the same for both cases (and add code comment to not return to it in the future)? Thanks in advance.

rlubos replied:
> I meant more of having accept be non-blocking, currently it blocks so you cannot really have a single loop do the accept / send / recv thing in parallel if I'm not mistaken. Does what I say makes sense?

@lbise You're right, the `accept` call seems to ignore `O_NONBLOCK` option, and always blocks waiting for an incoming connection (this should be an easy fix though). As an alternative, you could `poll` the listening socket, and call `accept` only when a new connection is present (i. e. POLLIN is returned).

pfalcon replied:
> You're right, the accept call seems to ignore O_NONBLOCK option, and always blocks waiting for an incoming connection (this should be an easy fix though).

So, is this defect of TLS implementation specifically? Because otherwise we have https://github.com/zephyrproject-rtos/zephyr/issues/6446 closed.

rlubos replied:
Correct, it was fixed for regular sockets in https://github.com/zephyrproject-rtos/zephyr/commit/0b1034aee20b15093c38399608186386d8ddd624

jukkar replied:
> Correct, it was fixed for regular sockets in [0b1034a](https://github.com/zephyrproject-rtos/zephyr/commit/0b1034aee20b15093c38399608186386d8ddd624)

Hmm, in that commit, I was actually returning EAGAIN which is probably not ok if the socket is blocking one. We need to fix that one.

rlubos commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -1277,6 +1277,11 @@ int ztls_accept_ctx(struct net_context *parent, struct sockaddr *addr,
 	}
 
 	child = k_fifo_get(&parent->accept_q, K_FOREVER);
+	if (child == NULL) {
```
Why would `k_fifo_get` return a NULL if it's called with `K_FOREVER` timeout? AFAIK it's not possible, unless there's some bug in `k_fifo_get`. @jukkar Or do I miss something?

jukkar replied:
The `k_fifo_get()` can return if `k_fifo_cancel_wait()` is called for it.

pfalcon reviewed (dismissed):
Let's please don't haste with this change, and a chance to figure out the right error code for this condition, as suggested in https://github.com/zephyrproject-rtos/zephyr/pull/26097#discussion_r437969229 .

This may save some head-scratching when porting existing (non-trivial) software to Zephyr.

To avoid premature merge before this is discussed, let me give -1 for now.

lbise commented:
> Let's please don't haste with this change, and a chance to figure out the right error code for this condition, as suggested in [#26097 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/26097#discussion_r437969229) .
> 
> This may save some head-scratching when porting existing (non-trivial) software to Zephyr.
> 
> To avoid premature merge before this is discussed, let me give -1 for now.

I've tested it on my linux machine by creating a simple program.

`#include <pthread.h>
#include <sys/socket.h>
#include <stdio.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>

pthread_t tid[2];
int listenfd = 0;

void* acceptThread(void *arg)
{
	int connfd = 0;
	char sendBuff[1025];
	struct sockaddr_in serv_addr;

	printf("acceptThread\n");

	listenfd = socket(AF_INET, SOCK_STREAM, 0);
	memset(&serv_addr, '0', sizeof(serv_addr));
    	memset(sendBuff, '0', sizeof(sendBuff));

    	serv_addr.sin_family = AF_INET;
    	serv_addr.sin_addr.s_addr = htonl(INADDR_ANY);
    	serv_addr.sin_port = htons(5000);

    	bind(listenfd, (struct sockaddr*)&serv_addr, sizeof(serv_addr));

    	listen(listenfd, 10);

	printf("Waiting for connections... fd = %d\n", listenfd);
	connfd = accept(listenfd, (struct sockaddr*)NULL, NULL);
	printf("Accept unblocked %d %d\n", connfd, errno);


}

void* closeThread(void *arg)
{
	int err;

	printf("closeThread fd = %d\n", listenfd);

	err = close(listenfd);
	if (err) {
		printf("Cannot close fd %d %d\n", err, errno);
	}

	printf("End closeThread\n");
}

int main(void)
{
	int err;
	void *res;

	printf("Start test...\n");

	err = pthread_create(&(tid[0]), NULL, &acceptThread, NULL);
	if (err) {
		printf("Cannot create accept thread\n");
		return -1;
	}

	sleep(5);

	err = pthread_create(&(tid[1]), NULL, &closeThread, NULL);
	if (err) {
		printf("Cannot create close thread\n");
		return -1;
	}

	err = pthread_join(tid[0], &res);

	return 0;
}
`
What happens is that when the second thread starts and tries to close the listening fd nothing happens. close returns 0 but the accept thread stays blocked.
In order to achieve what I was looking for I had to replace the close call by shutdown. In that case the accept thread is unblocked and returns EINAL.
`Start test...
acceptThread
Waiting for connections... fd = 3
closeThread fd = 3
End closeThread
Accept unblocked -1 22
`

lbise committed:
net: sockets: Return EINVAL when an accept call is unblocked

Under Linux when you shutdown a socket which is blocked on
an accept call the error code returned by accept is EINVAL.
Modify the socket code to be inline with this behaviour.

Signed-off-by: Léonard Bise <leonard.bise@gmail.com>

lbise commented:
> @lbise: Thanks for taking time to write that testcase!
> 
> > What happens is that when the second thread starts and tries to close the listening fd nothing happens. close returns 0 but the accept thread stays blocked.
> 
> Continuing from that, when you make a connection, accept returns a valid client socket. And when you call accept() again, it returns EBADF, which is well expected in this case. The behavior of "close() doesn't have effect on pending accept()" can be explained by the fact that in Linux, underlying socket objects are reference-counted. So, even if we release one fd pointing to underlying socket, accept() still keeps a reference internally, and so the socket object isn't closed until accept() returns.
> 
> As we don't use refcounting, behaving like shutdown() makes sense, and that would be returning EINVAL as you point out.
> 
> So (if everyone else agrees to EINVAL), could you make a second commit changing subsys/net/lib/sockets/sockets.c to EINVAL too? And with code comment in both files (in addition to commit message) describing the situation, so we didn't repeat this exercised in the future. Thanks in advance!

I updated as you suggested. Let me know if comments are good enough.

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -465,7 +465,10 @@ int zsock_accept_ctx(struct net_context *parent, struct sockaddr *addr,
+		/* Return EINVAL which is the same error code used under Linux
+		 * when calling shutdown on a blocked accept call
+		 */
+		errno = EINVAL;
```
Hmm, but if the timeout was K_NO_WAIT, then the EAGAIN would be appropriate here.

lbise replied:
Good point, I added a condition
